std/assembly/array.ts

```diff
@@ -1,148 +1,156 @@
+    return this.__unchecked_get(index);
   }
 
-  @operator("{}")
-  private __unchecked_get(index: i32): T {
-    return LOAD<T>(this.buffer_, index);
+  @operator("{}") private __unchecked_get(index: i32): T {
+    return load<T>(this.dataStart + (<usize>index << alignof<T>()));
   }
 
-  @operator("[]=")
-  private __set(index: i32, value: T): void {
-    var buffer = this.buffer_;
-    var capacity = buffer.byteLength >>> alignof<T>();
-    if (<u32>index >= <u32>capacity) {
-      const MAX_LENGTH = MAX_BLENGTH >>> alignof<T>();
-      if (<u32>index >= <u32>MAX_LENGTH) throw new Error("Invalid array length");
-      buffer = reallocateUnsafe(buffer, (index + 1) << alignof<T>());
-      this.buffer_ = buffer;
-      this.length_ = index + 1;
+  @operator("[]=") private __set(index: i32, value: T): void {
+    var length = this.length_;
+    if (isReference<T>()) {
+      if (!isNullable<T>()) {
+        if (<u32>index > <u32>length) throw new Error(E_HOLEYARRAY);
+      }
     }
-    STORE<T>(buffer, index, value);
-    if (isManaged<T>()) __gc_link(changetype<usize>(this), changetype<usize>(value)); // tslint:disable-line
+    ensureSize(changetype<usize>(this), index + 1, alignof<T>());
+    this.__unchecked_set(index, value);
+    if (index >= length) this.length_ = index + 1;
   }
 
-  @operator("{}=")
-  private __unchecked_set(index: i32, value: T): void {
-    STORE<T>(this.buffer_, index, value);
-    if (isManaged<T>()) __gc_link(changetype<usize>(this), changetype<usize>(value)); // tslint:disable-line
+  @operator("{}=") private __unchecked_set(index: i32, value: T): void {
+    if (isManaged<T>()) {
+      let offset = this.dataStart + (<usize>index << alignof<T>());
+      let oldRef: usize = load<usize>(offset);
+      if (changetype<usize>(value) != oldRef) {
+        store<usize>(offset, __retain(changetype<usize>(value)));
+        __release(oldRef);
+      }
+    } else {
+      store<T>(this.dataStart + (<usize>index << alignof<T>()), value);
+    }
   }
 
   fill(value: T, start: i32 = 0, end: i32 = i32.MAX_VALUE): this {
-    var buffer = this.buffer_;
-    var len    = this.length_;
-
-    start = start < 0 ? max(len + start, 0) : min(start, len);
-    end   = end   < 0 ? max(len + end,   0) : min(end,   len);
-
+    var dataStart = this.dataStart;
+    var length = this.length_;
+    start = start < 0 ? max(length + start, 0) : min(start, length);
+    end   = end   < 0 ? max(length + end,   0) : min(end,   length);
     if (sizeof<T>() == 1) {
       if (start < end) {
         memory.fill(
-          changetype<usize>(buffer) + start + HEADER_SIZE,
-          <u8>value,
+          dataStart + <usize>start,
+          u8(value),
           <usize>(end - start)
         );
       }
     } else {
       for (; start < end; ++start) {
-        STORE<T>(buffer, start, value);
+        store<T>(dataStart + (<usize>start << alignof<T>()), value);
```

Comment:
I guess this need check `isManged<T>()` and insert `__gc_link`. wdyt?

Reply:
Good find, same seems to be the case for copyWithin